traffic_router/core/src/main/java/org/apache/traffic_control/traffic_router/core/dns/NameServer.java

```diff
@@ -328,11 +329,19 @@ private static RRset setNegativeTTL(final RRset original, final int flags) {
 			if (record instanceof SOARecord) {
 				final SOARecord soa = (SOARecord) record;
 
+				// Set the "minimum" attribute to be the maximum of the current minimum value and 900 (seconds)
+				// This is done to increase the negative caching TTL, so as to maximize the time interval between
+				// successive NXDOMAIN or NXRRSET responses.
+				final long minimum = Math.max(soa.getMinimum(), negativeCachingTTL);
+				record = new SOARecord(soa.getName(), DClass.IN, soa.getTTL(), soa.getHost(), soa.getAdmin(),
+						soa.getSerial(), soa.getRefresh(), soa.getRetry(), soa.getExpire(),
+						minimum);
+
 				// the value of the minimum field is less than the actual TTL; adjust
 				if (soa.getMinimum() != 0 || soa.getTTL() > soa.getMinimum()) {
 					record = new SOARecord(soa.getName(), DClass.IN, soa.getMinimum(), soa.getHost(), soa.getAdmin(),
 							soa.getSerial(), soa.getRefresh(), soa.getRetry(), soa.getExpire(),
-							soa.getMinimum());
+							minimum);
 				} // else use the unmodified record
```

Comment:
Instead of potentially setting `record` twice here, use a condition for `ttl` field and set `record` only once?

Reply:
I was actually going to do that, but then I realized that `record` doesn't have a setter method for TTL, so I have to redeclare the complete value.

Reply:
> I was actually going to do that, but then I realized that `record` doesn't have a setter method for TTL, so I have to redeclare the complete value.

I mean setting the `ttl` argument around that logic before computing the whole `record`. Like:

```java
				final long minimum = Math.max(soa.getMinimum(), negativeCachingTTL);
				final long ttl = soa.getMinimum() != 0 || soa.getTTL() > soa.getMinimum()
						// the value of the minimum field is less than the actual TTL; adjust
						? soa.getMinimum()
						// else use the unmodified record
						: soa.getTTL();
				record = new SOARecord(soa.getName(), DClass.IN, ttl, soa.getHost(), soa.getAdmin(),
						soa.getSerial(), soa.getRefresh(), soa.getRetry(), soa.getExpire(),
						minimum);
```

or so?